File: JVM/src/main/kotlin/io/alliance/adn/syntactic/Parser.kt
package io.alliance.adn.syntactic

import io.alliance.adn.data.ElementArray
import io.alliance.adn.data.ElementLiteral
import io.alliance.adn.data.ElementRoot
import io.alliance.adn.data.ElementStruct
import io.alliance.adn.data.ElementType
import io.alliance.adn.Walkable
import io.alliance.adn.api.InvalidCastException
import io.alliance.adn.api.InvalidTokenException
import io.alliance.adn.data.*
import io.alliance.adn.semantic.Token
import io.alliance.adn.semantic.TokenIdentifier
import io.alliance.adn.semantic.TokenKind
import io.alliance.adn.semantic.TokenKind.*
import io.alliance.adn.semantic.TokenLiteral
import io.alliance.adn.semantic.TokenType

internal class Parser(input: List<Token>) : Walkable<Token>(input) {
    private fun match(kind: TokenKind): Token {
        if (current kindOf kind) {
            return consume
        }

        throw InvalidTokenException("Unexpected token <$current>, expected <$kind>")
    }

    private fun matchIdentifier(): TokenIdentifier {
        if (current kindOf IDENTIFIER) {
            return consume as TokenIdentifier
        }

        throw InvalidTokenException("Unexpected token <$current>, expected primitive type keyword")
    }

    private fun matchType(): TokenType {
        if (current.kind.isPrimitive) {
            return consume as TokenType
        }

        throw InvalidTokenException("Unexpected token <$current>, expected primitive type keyword")
    }

    private fun matchLiteral(): TokenLiteral {
        if (current.kind.isLiteral) {
            return consume as TokenLiteral
        }

        throw InvalidTokenException("Unexpected token <$current>, expected literal value")
    }

    internal fun parse(): ElementRoot {
        val root = ElementRoot()

        while (!(current kindOf MISC_END_OF_FILE)) {
            root += parseElement()
        }

        return root
    }

    private fun parseElement(): Element {
        // Sort out structs first

        // struct MyStruct {
        if (current kindOf KEYWORD_STRUCT) {
            consume
            return parseStruct(matchIdentifier())
        }

        val identifier = matchIdentifier()

        // MyStruct {
        if (current kindOf TOKEN_BRACE_OPEN) {
            return parseStruct(identifier)
        }

        // Structs are sorted out, now let's deal with types

        // myNumber: u32 = 0
        // MyArray: u32[][] [ [
        if (current kindOf TOKEN_COLON) {
            consume

            val type = matchType()

            var depth = 0
            while (current kindOf KEYWORD_ARRAY) {
                consume
                depth++
            }

            if (depth > 0) {
                return parseArray(identifier, type.asType(), depth)
            }

            return parseLiteral(identifier, type)
        }

        // myNumber 0
        // MyArray [ [

        var depth = 0
        while (peek(depth) kindOf TOKEN_BRACKET_OPEN) {
            depth++
        }

        if (depth > 0) {
            return parseArray(identifier, ElementType.NULL, depth)
        }

        return parseLiteral(identifier, null)
    }

    private fun parseStruct(identifier: TokenIdentifier): Element {
        match(TOKEN_BRACE_OPEN)

        val struct = ElementStruct(identifier.name)
        while (!(current kindOf TOKEN_BRACE_CLOSE)) {
            struct += parseElement()
        }

        match(TOKEN_BRACE_CLOSE)
        return struct
    }

    private fun parseArray(identifier: TokenIdentifier, expectedType: ElementType, depth: Int): Element {
        // <MyArray> [ [ 2, 9 ], [
        val (actualType, elements) = parseArrayElement()

        if (expectedType == ElementType.NULL) {
            return mapArray(identifier.name, elements, actualType, -(1 - depth), 0)
        }

        if (expectedType != expectedType * actualType) {
            throw InvalidCastException(identifier.name, actualType, expectedType)
        }

        return mapArray(identifier.name, elements, expectedType, -(1 - depth), 0)
    }

    private fun mapArray(name: String, list: ArrayList<*>, type: ElementType, depth: Int, layer: Int): ElementArray {
        if (depth != layer) {
            val array = ElementArray(name, type, -(layer - depth))

            for (nestedList in list) {
                array += mapArray(name, nestedList as ArrayList<*>, type, depth, layer + 1)
            }

            return array
        }

        val array = ElementArray(name, type, 0)

        // parseArrayElement cannot produce a list of anything else on the deepest nesting level.
        @Suppress("UNCHECKED_CAST")
        for (value in list as ArrayList<TokenLiteral>) {
            array += ElementLiteral.anonymous(value.value, type)
        }

        return array
    }

    private fun parseArrayElement(): Pair<ElementType, ArrayList<*>> {
        var type = ElementType.NULL

        match(TOKEN_BRACKET_OPEN)

        // 'MyArray [ [ ... ], ... ]'
        if (current kindOf TOKEN_BRACKET_OPEN) {
            val list = ArrayList<ArrayList<*>>()

            while (!(current kindOf TOKEN_BRACKET_CLOSE)) {
                val (t, l) = parseArrayElement()
                type *= t
                list.add(l)

                // comma can be omitted
                if (current kindOf TOKEN_COMMA) {
                    consume
                }
            }

            consume
            return Pair(type, list)
        }

        val list = ArrayList<TokenLiteral>()

        // 'MyArray [ 0, ... ]'
        while (!(current kindOf TOKEN_BRACKET_CLOSE)) {
            val literal = matchLiteral()
            type *= inferType(literal)
            list.add(literal)

            // comma can be omitted
            if (current kindOf TOKEN_COMMA) {
                consume
            }
        }

        consume
        return Pair(type, list)
    }

    private fun parseLiteral(identifier: TokenIdentifier, maybeType: TokenType?): ElementLiteral<*> {
        // value: u32 = 0;
        // Equals and semicolon may be omitted

        if (current kindOf TOKEN_SET) {
            consume
        }

        val value = matchLiteral()
        val type = maybeType?.kind ?: inferKind(value.kind)

        if (current kindOf TOKEN_SEMICOLON) {
            consume
        }

        return ElementLiteral.create(identifier, type, value)
    }

    private fun inferKind(kind: TokenKind): TokenKind {
        return when (kind) {
            LITERAL_TRUE, LITERAL_FALSE -> KEYWORD_BOOL
            LITERAL_NUMBER -> KEYWORD_I32
            LITERAL_FLOAT -> KEYWORD_F32
            LITERAL_STRING -> KEYWORD_STR
            else -> throw InvalidTokenException("Unexpected token <$current>, expected literal")
        }
    }

    private fun inferType(token: TokenLiteral): ElementType {
        return when (token.kind) {
            LITERAL_TRUE, LITERAL_FALSE -> ElementType.BOOL
            LITERAL_NUMBER -> ElementType.I32
            LITERAL_FLOAT -> ElementType.F32
            LITERAL_STRING -> ElementType.STR
            else -> throw InvalidTokenException("Unexpected token <$current>, expected literal")
        }
    }
}

File: Native/src/syntactic/parser.rs
use std::cell::Cell;
use std::collections::HashMap;
use std::fmt::Error;

use crate::data::{Datapoint, Dataset};
use crate::semantic::TokenKind;
use crate::semantic::TokenKind::*;
use crate::Walkable;

macro_rules! match_token {
    ($self:ident $(,)? $token:pat) => {
        match $self.current() {
            $token => (),
            _ => return Err(Error::default()),
        }
    };
}

pub(crate) struct Parser {
    tokens: Vec<TokenKind>,
    index: Cell<usize>,
}

impl Parser {
    pub(crate) fn parse(tokens: Vec<TokenKind>) -> Result<Dataset, Error> {
        let parser = Parser {
            tokens,
            index: Cell::new(0),
        };

        let mut root = Dataset::new();

        while !matches!(parser.current(), MiscEndOfFile) {
            let (name, data) = parser.parse_element()?;
            root.add(name, data);
        }

        Ok(root)
    }

    fn parse_element(&self) -> Result<(String, Datapoint), Error> {
        // Sort out structs first

        // struct MyStruct {
        if matches!(self.current(), KeywordStruct) {
            self.skip();
            return match self.consume() {
                Identifier(name) => Ok((name.clone(), self.parse_struct()?)),
                _ => Err(Error::default()),
            }
        }

        let identifier = match self.consume() {
            Identifier(name) => name.clone(),
            _ => return Err(Error::default()),
        };

        // MyStruct {
        if matches!(self.current(), TokenBraceOpen) {
            return Ok((identifier, self.parse_struct()?));
        }

        // Structs are sorted out, now let's deal with types

        // myNumber: u32 = 0
        // MyArray: u32[][] [ [
        if matches!(self.current(), TokenColon) {
            self.skip();

            let cur = self.current();
            let _type = match cur {
                KeywordI8 | KeywordI16 | KeywordI32 | KeywordI64 |
                KeywordBool | KeywordF32 | KeywordF64 | KeywordStr => cur,
                _ => return Err(Error::default())
            };

            let mut depth: usize = 0;
            while matches!(self.current(), KeywordArray) {
                self.skip();
                depth += 1;
            }

            return match depth {
                0 => Ok((identifier, self.parse_literal(Some(_type))?)),
                _ => Ok((identifier, self.parse_array(Some(_type), depth as i32)?))
            }
        }

        // Type annotations have been dealt with, no parse inferred values

        // myNumber 0
        // MyArray [ [
        let mut depth: usize = 0;
        while matches!(self.peek(depth), TokenBracketOpen) {
            depth += 1;
        }

        return match depth {
            0 => Ok((identifier, self.parse_literal(None)?)),
            _ => Ok((identifier, self.parse_array(None, depth as i32)?))
        }
    }

    fn parse_struct(&self) -> Result<Datapoint, Error> {
        match_token!(self, TokenBraceOpen);

        let mut map: HashMap<String, Datapoint> = HashMap::new();
        while !matches!(self.current(), TokenBraceClose) {
            let (name, data) = self.parse_element()?;
            map.insert(name, data);
        }

        match_token!(self, TokenBraceClose);
        Ok(Datapoint::Struct(map))
    }

    fn parse_array(&self, expected_type: Option<&TokenKind>, depth: i32) -> Result<Datapoint, Error> {
        let (actual_type, elements): (&TokenKind, _) = self.parse_array_elements();

        if let Some(expected) = actual_type.castable_from(expected_type) {
            self.map_array(elements, expected, -(1 - depth), 0)
        } else {
            Err(Error::default())
        }
    }

    fn parse_literal(&self, expected_type: Option<&TokenKind>) -> Result<Datapoint, Error> {
        todo!()
    }
}

impl Walkable<TokenKind> for Parser {
    #[inline]
    fn size(&self) -> usize {
        self.tokens.len()
    }

    #[inline]
    fn peek(&self, offset: usize) -> &TokenKind {
        self.tokens.get(self.index.get() + offset).expect("Index out of bounds!")
    }

    #[inline]
    fn skip(&self) {
        self.index.set(self.index.get() + 1);
    }
}